### PR TITLE
Patch/pinned init v3

### PR DESCRIPTION
**Pull request by BennoLossin**

### CODE REVIEW

File: rust/macros/pin_data.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use proc_macro::{Punct, Spacing, TokenStream, TokenTree};
+
+pub(crate) fn pin_data(args: TokenStream, input: TokenStream) -> TokenStream {
+    // This proc-macro only does some pre-parsing and then delegates the actual parsing to
+    // `kernel::_pin_data!`.
```
Comment (metaspace) [inline rust/macros/pin_data.rs]: There is a typo here, should be `__pin_data`, right?

Comment (BennoLossin) [reply rust/macros/pin_data.rs]: That is correct, good catch!

File: rust/macros/pin_data.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    // `impl_generics`, the declared generics with their bounds.
+    let mut impl_generics = vec![];
+    // Only the names of the generics, without any bounds.
+    let mut ty_generics = vec![];
+    // Tokens not related to the generics e.g. the `impl` token.
+    let mut rest = vec![];
+    // The current level of `<`.
+    let mut nesting = 0;
+    let mut toks = input.into_iter();
+    // if we are at the beginning of a generic parameter
```
Comment (ojeda) [inline rust/macros/pin_data.rs]:
```suggestion
    // If we are at the beginning of a generic parameter.
```

File: rust/macros/lib.rs [hunk 3]
```diff
@@ -166,3 +170,79 @@ pub fn vtable(attr: TokenStream, ts: TokenStream) -> TokenStream {
...
+/// #[pinned_drop]
+/// impl PinnedDrop for DriverData {
+///     fn drop(self: Pin<&mut Self>) {
+///         unsafe { bindings::destroy_info(self.raw_info) };
+///     }
+/// }
+/// ```
+///
+/// [`pin_init!`]: ../kernel/macro.pin_init.html
+//  ^ cannot use direct link, since `kernel` is not a dependency of `macros`
```
Comment (ojeda) [inline rust/macros/lib.rs]:
```suggestion
//  ^ cannot use direct link, since `kernel` is not a dependency of `macros`.
```

File: rust/macros/lib.rs [hunk 4]
```diff
@@ -166,3 +170,79 @@ pub fn vtable(attr: TokenStream, ts: TokenStream) -> TokenStream {
...
 }
+
+/// Used to specify the pinning information of the fields of a struct.
+///
+/// This is somewhat similar in purpose as
+/// [pin-project-lite](https://crates.io/crates/pin-project-lite).
+/// Place this macro on a struct definition and then `#[pin]` in front of the attributes of each
+/// field you want to structurally pin.
+///
+/// This macro enables the use of the [`pin_init!`] macro. When pinned-initializing a `struct`,
```
Comment (ojeda) [inline rust/macros/lib.rs]: Should it be "pin-initializing"?

File: rust/kernel/sync/arc.rs [hunk 5]
```diff
@@ -163,6 +167,28 @@ impl<T> Arc<T> {
...
+    pub fn pin_init<E>(init: impl PinInit<T, E>) -> Result<Self>
+    where
+        Error: From<E>,
+    {
+        UniqueArc::pin_init(init).map(|u| u.into())
+    }
+
+    /// Use the given initializer to in-place initialize a `T`.
+    ///
+    /// This is equivalent to [`pin_init`], since an `Arc` is always pinned.
```
Comment (ojeda) [inline rust/kernel/sync/arc.rs]: Can the `Arc` be an intra-doc link?

File: rust/kernel/init/macros.rs [hunk 6]
```diff
@@ -0,0 +1,481 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module provides the macros that actually implement the proc-macros `pin_data` and
+//! `pinned_drop`. These macros should never be called directly, since they expect their input to be
+//! in a certain format which is internal. Use the proc-macros instead.
+//!
+//! This architecture has been chosen because the kernel does not yet have access to `syn` which
+//! would make matters a lot easier for implementing these as proc-macros.
+
+/// This macro creates a `unsafe impl<...> PinnedDrop for $type` block.
```
Comment (ojeda) [inline rust/kernel/init/macros.rs]:
```suggestion
/// Creates a `unsafe impl<...> PinnedDrop for $type` block.
```

File: rust/kernel/init/macros.rs [hunk 7]
```diff
@@ -0,0 +1,481 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module provides the macros that actually implement the proc-macros `pin_data` and
```
Comment (ojeda) [inline rust/kernel/init/macros.rs:3]: Typically we have a single line with the title ("short description"). It does not matter much since this is `doc(hidden)`, but still, it would be nice to have it.

**Comment by BennoLossin** [comment]
for anyone wanting to see the individual commits, see https://github.com/y86-dev/linux/tree/patch/pinned-init-v3-not-squashed

File: rust/kernel/lib.rs [hunk 8]
```diff
@@ -16,7 +16,9 @@
 #![feature(coerce_unsized)]
 #![feature(core_ffi_c)]
 #![feature(dispatch_from_dyn)]
+#![feature(explicit_generic_args_with_impl_trait)]
 #![feature(generic_associated_types)]
+#![feature(new_uninit)]
```
Comment (BennoLossin) [inline rust/kernel/lib.rs:21]: should these feature enables be in their own commits?

File: scripts/Makefile.build [hunk 9]
```diff
@@ -277,7 +277,7 @@ $(obj)/%.lst: $(src)/%.c FORCE
 # Compile Rust sources (.rs)
 # ---------------------------------------------------------------------------
 
-rust_allowed_features := core_ffi_c
+rust_allowed_features := core_ffi_c,explicit_generic_args_with_impl_trait
```
Comment (BennoLossin) [inline scripts/Makefile.build:280]: Similarly this?

This feature was stabilized in 1.63, so this might not be here for long.

File: rust/kernel/init.rs [hunk 10]
```diff
@@ -0,0 +1,1337 @@
...
+        );
+    };
+    (init_slot:
+        @data($data:ident),
+        @slot($slot:ident),
+        // Direct value init, this is safe for every field.
+        @munch_fields($field:ident $(: $val:expr)?, $($rest:tt)*),
+    ) => {
+        $(let $field = $val;)?
+        // Call the initializer.
```
Comment (metaspace) [inline rust/kernel/init.rs]: Should this be "Initialize the field" instead maybe? Since no initializer is actually called in this case?

**Comment by metaspace** [comment]
I think we should split up `rust: add pin-init API` in multiple patches. I realize that many of the types used by the macros are not meant to be used by consumers of this API, but I think this would be easier to review if the types used by the macros were added on their own, along with examples of intended use. The macros could then be added in a separate patch as a means to automate the examples.

**Comment by BennoLossin** [comment]
> I think we should split up `rust: add pin-init API` in multiple patches. I realize that many of the types used by the macros are not meant to be used by consumers of this API, but I think this would be easier to review if the types used by the macros were added on their own, along with examples of intended use. The macros could then be added in a separate patch as a means to automate the examples.

Not sure if I understand correctly, are you talking about the types inside of `__internal.rs`, or the other types as well? What do you mean by examples of intended use? The `__internal.rs` types mostly serve a single purpose and are only intended to be used inside of the macros. I could give some macro expanded code as examples of how the macro uses these internal types if that is what you mean, but I dont know where to put them for the patches.

**Comment by metaspace** [comment]
> Not sure if I understand correctly, are you talking about the types inside of `__internal.rs`, or the other types as well? What do you mean by examples of intended use? The `__internal.rs` types mostly serve a single purpose and are only intended to be used inside of the macros. I could give some macro expanded code as examples of how the macro uses these internal types if that is what you mean, but I dont know where to put them for the patches.

I think the `rust: add pin-init API` commit is too big. For me it was difficult to digest, so I think a bit more help to the reader would be nice. I would suggest the following:

 * Splitting the commit in at least two: One part that supports `pin_init_from_closure` and one part that adds macros that create the initializer
 * Even though `__internal` is not intended to be used by consumers of the `init` API, it is a public module that is going to be accessible to the user, right? So the user could write something like the expanded macro text, calling into `init::__internal`. Adding `__internal` in one commit along with examples of how the expanded macros will use it would be a major help in understanding the macros. The examples could be delivered in the module documentation.
 * I would suggest splitting in even more pieces, like one adding `init!` and one adding `pin_init!`
 * The commit message is nice. I would love to have that as module documentation. I think the commit message is better than the module level intro.
 * It was difficult for me to parse the macros. I am not a macro expert, I had to look up quite a bit of mechanics to follow along. One thing that would have been a great help is having expanded examples in the documentation. It was difficult for me to read the macros without knowing where it was trying to end up. I know there are tools for looking at expanded macros, but I still think some small examples in the documentation would be nice.
 * I did not immediately understand some of the traits that are used to help type inference. It would be great to document why and how this works in the expanded macro examples.

I hope this makes more sense :) I am not sure who is going to review this in the end, but if it is people that are not long time Rust experts, we need to give them all the help we can think of.

**Comment by nbdd0121** [comment]
Types used only internally by the macro should *not* be a separate commit. Each logical change should be a commit, no matter how big or small it is.

Splitting the commit into a core `PinInit/Init/StackInit/pin_init_from_closure` commit and a macro commit, though, does make sense.

File: rust/kernel/init/macros.rs [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+        // We prevent this by creating a trait that will be implemented for all types implementing
+        // `Drop`. Additionally we will implement this trait for the struct leading to a conflict,
+        // if it also implements `Drop`
+        trait MustNotImplDrop {}
+        #[allow(drop_bounds)]
+        impl<T: ::core::ops::Drop> MustNotImplDrop for T {}
+        impl<$($impl_generics)*> MustNotImplDrop for $name<$($ty_generics)*>
+        where $($whr)* {}
+        // We also take care to prevent users from writing a useless PinnedDrop implementation.
+        // They might implement PinnedDrop correctly for the struct, but forget to give
```
Comment (ojeda) [inline rust/kernel/init/macros.rs]:
```suggestion
        // We also take care to prevent users from writing a useless `PinnedDrop` implementation.
        // They might implement `PinnedDrop` correctly for the struct, but forget to give
```

File: rust/kernel/init/macros.rs [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+        // `PinnedDrop` as the parameter to `#[pin_data]`.
+        #[allow(non_camel_case_types)]
+        trait UselessPinnedDropImpl_you_need_to_specify_PinnedDrop {}
+        impl<T: $crate::init::PinnedDrop>
+            UselessPinnedDropImpl_you_need_to_specify_PinnedDrop for T {}
+        impl<$($impl_generics)*>
+            UselessPinnedDropImpl_you_need_to_specify_PinnedDrop for $name<$($ty_generics)*>
+        where $($whr)* {}
+    };
+    // When `PinnedDrop` was specified we just implement drop and delegate.
```
Comment (ojeda) [inline rust/kernel/init/macros.rs]:
```suggestion
    // When `PinnedDrop` was specified we just implement `Drop` and delegate.
```

File: rust/kernel/init/macros.rs [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+        @impl_generics($($impl_generics:tt)*),
+        @ty_generics($($ty_generics:tt)*),
+        @where($($whr:tt)*),
+        @pinned_drop(PinnedDrop),
+    ) => {
+        impl<$($impl_generics)*> ::core::ops::Drop for $name<$($ty_generics)*>
+        where $($whr)*
+        {
+            fn drop(&mut self) {
+                // SAFETY: since this is a destructor, `self` will not move after this function
```
Comment (ojeda) [inline rust/kernel/init/macros.rs]:
```suggestion
                // SAFETY: Since this is a destructor, `self` will not move after this function
```

File: rust/kernel/init/macros.rs [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+                // terminates, since it is inaccessible.
+                let pinned = unsafe { ::core::pin::Pin::new_unchecked(self) };
+                // SAFETY: since this is a drop function, we can create this token to call the
```
Comment (ojeda) [inline rust/kernel/init/macros.rs]:
```suggestion
                // SAFETY: Since this is a drop function, we can create this token to call the
```

**Review by ojeda — COMMENTED** [review]
Some trivial comments from last week from when the GH outage happened, please ignore any if they are not relevant after the later push(es).

File: scripts/Makefile.build (hunk 9 above)
Comment (ojeda) [reply scripts/Makefile.build:280]: Don't worry about it -- Lina has a patch enabling it too, so I can pick that one first: https://lore.kernel.org/rust-for-linux/20230224-rust-new_uninit-v1-1-c951443d9e26@asahilina.net/

File: rust/kernel/lib.rs (hunk 8 above)
Comment (ojeda) [reply rust/kernel/lib.rs:21]: Don't worry about it -- Lina has a patch enabling it too, so I can pick that one first: https://lore.kernel.org/rust-for-linux/20230224-rust-new_uninit-v1-1-c951443d9e26@asahilina.net/

File: scripts/Makefile.build (hunk 9 above)
Comment (ojeda) [reply scripts/Makefile.build:280]: No worries either, it will go away in the merge cycle.

**Comment by ojeda** [comment]
> Each logical change should be a commit, no matter how big or small it is.

This is true, and it simplifies reverting changes later and so on. However, big patches may be split into smaller ones to make things more understandable for reviewers, even if they could also be understood to go together as a unit too, as long as the series does not break anything after each step (so fairly easy when introducing something new).

Having said that, 2k lines is still within reason (even if already within the top percentile) and several versions have been in the list for a while, but if you (@y86-dev) can do a logical split somewhere that makes it easier to understand (taking into account the commit messages too, especially given the one you have there), please go for it for v4 since some people will appreciate it (not just now, but in the future too).

**Comment by ojeda** [comment]
GitHub's "Compare" button says `from_value` is gone.

Also "rust: add pin-init API items to prelude" -> "rust: prelude: add pin-init API items".

**Comment by BennoLossin** [comment]
> GitHub's "Compare" button says `from_value` is gone.
> 
> Also "rust: add pin-init API items to prelude" -> "rust: prelude: add pin-init API items".

Yeah that was replaced by `impl Init<T> for T`, had forgotten to remove it when I added that. Also some other things are not completely done, still rebasing stuff.

**Comment by ojeda** [comment]
I saw it (`from_value`) referenced from intra-doc links too.

**Comment by BennoLossin** [comment]
@metaspace I added a macro expansion example, let me know what you think of it.

Not sure if the `__internal.rs` docs is the best place for it, since that is `#[doc(hidden)]`, maybe I could create a `doc` module just for that purpose? Also not sure into which commit this should go, since `PinnedDrop` is now split from `pin_init!`.

**Comment by metaspace** [comment]
> @metaspace I added a macro expansion example, let me know what you think of it.
> 
> Not sure if the `__internal.rs` docs is the best place for it, since that is `#[doc(hidden)]`, maybe I could create a `doc` module just for that purpose? Also not sure into which commit this should go, since `PinnedDrop` is now split from `pin_init!`.

I think this is really helpful. I think having it where you put it is fine. It's not something a user would read anyway.

File: rust/kernel/init/macros.rs [hunk 15]
```diff
@@ -0,0 +1,964 @@
...
+//! let a = 42;
+//! let initializer = pin_init!(Foo {
+//!     a,
+//!     b <- Bar::new(36),
+//! });
+//! ```
+//!
+//! This example includes the most common and important features of the pin-init API.
+//!
+//! Below you can find induvidual section about the different macro invocations. Here are some
```
Comment (ojeda) [inline rust/kernel/init/macros.rs]:
```suggestion
//! Below you can find individual section about the different macro invocations. Here are some
```

File: rust/kernel/sync/arc.rs [hunk 16]
```diff
@@ -540,6 +540,31 @@ impl<T> UniqueArc<MaybeUninit<T>> {
...
+    pub fn pin_init_with<E>(
+        mut self,
+        init: impl PinInit<T, E>,
+    ) -> core::result::Result<Pin<UniqueArc<T>>, E> {
+        // SAFETY: The supplied pointer is valid for initialization and we will later pin the value
+        // to ensure it does not move.
+        match unsafe { init.__pinned_init(self.as_mut_ptr()) } {
+            // SAFETY: Initialization completed successfully and we can pin the value, since this
+            // is an owning pointer.
+            Ok(()) => Ok(unsafe { Pin::new_unchecked(self.assume_init()) }),
```
Comment (fbq) [inline rust/kernel/sync/arc.rs]: Maybe add a `UniqueArc::into_pin`, similar to `Box::into_pin`?

Comment (BennoLossin) [reply rust/kernel/sync/arc.rs]: Oh I did not see that we already have `From<UniqueArc<T>> for Pin<UniqueArc<T>>`, I can just use that.

**Comment by BennoLossin** [comment]
continued at #991
